### PR TITLE
Adding file persistence bundle and endpoint handling multipart data types

### PR DESCRIPTION
**Pull request by VanDavv**
So, finally I made it to the end. Whole file persistence actions are put into one bundle, I also made it flexible using interfaces, so that migration to Windows Azure wont be a major problem. 

Only issue I had was Authorization. I analysed the code and found out, that CSV file with initial data is not wrote correctly, because there shouldn't be any whitespaces between values. As it follows, my IntegrationTest was using wrong credentials at first, so I got 400 Bad Request status code, but then I added whitespaces, but still, response code was 401, not 201. But, when I turned off authorization, whole bundle worked just fine, so I think that this PR makes sense.

During implementation process I found some issues with CV:

- There is no working api or even mock!
- They're returning flashcards collection as a response, but those flashcards have included id's. From where they get these id's?
- After the flashcards collection, there is a status code (0, 1 or 2). As we're using HTTP, can't they just return an appropriate status code for the response? It would be much easier to read such an answer
- When posting a request, we need to specify type of method we want to use ("ImageToFlashcard" etc). But, during whole persisting process, we're unaware of filetype that we're dealing with. Couldn't CV have the action recognition implemented on their serwer? 

For now, there are commented places in streamPersistenceBundle, which need a real CV server to work with.

### CODE REVIEW

File: backend/app/pom.xml [hunk 1]
```diff
@@ -26,6 +26,11 @@
             <groupId>de.svenkubiak</groupId>
             <artifactId>jBCrypt</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.bls.patronage</groupId>
+            <artifactId>streamPersistenceBundle</artifactId>
+            <version>1.0-SNAPSHOT</version>
```
Comment (mlotysz) [inline backend/app/pom.xml]: use `${project.version}`

File: backend/app/src/main/java/com/bls/patronage/StudyBox.java [hunk 2]
```diff
@@ -37,6 +39,14 @@
     private static final String APP_NAME = "backend";
     private static final String HEALTH_CHECK_DATABASE_NAME = "database";
 
+    private final StreamPersistenceBundle<StudyBoxConfiguration> streamPersistenceBundle = new StreamPersistenceBundle<StudyBoxConfiguration>() {
+        @Override
+        public Listener getListener(StudyBoxConfiguration configuration) {
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/StudyBox.java]: Wrong name. It's not an listener. It's configuration.

File: backend/app/src/main/java/com/bls/patronage/StudyBox.java [hunk 3]
```diff
@@ -37,6 +39,14 @@
...
+            //Return listener with CV server URI
+            return () -> configuration.getCvServerURI();
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/StudyBox.java]: remove ().

Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/StudyBox.java]: see other DW bundles how configuration is done. is much easier than your solution.

File: backend/app/src/main/java/com/bls/patronage/auth/PreAuthenticationFilter.java [hunk 4]
```diff
@@ -26,9 +27,15 @@ public void filter(ContainerRequestContext request) throws IOException {
             }
         }
         if (!hasValidHeader) {
-            if (request.getMethod() != "GET") {
+            final boolean isPostingUser =
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/auth/PreAuthenticationFilter.java]: Why this is changed here? Rebase onto new master.

File: backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        this.baseLocation = baseLocation;
+        this.flashcardDAO = flashcardDAO;
+    }
+
+    @POST
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response uploadFile(
+            @Auth User user,
+            @FormDataParam("file") InputStream uploadedInputStream,
+            @FormDataParam("file") FormDataContentDisposition details) throws Exception {
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: `details` is not used.

File: backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        final Deck deck = new Deck(UUID.randomUUID(), "foo");
+            final java.nio.file.Path location = baseLocation.resolve(user.getId().toString());
+
+
+        final URL streamedFileLocation = bundle.persistStream(uploadedInputStream, location.toUri().toURL());
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: This class shouldn't know bundle. Some service interface is preferred.

File: backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        final Response response = bundle.informListener(new CVMessage(streamedFileLocation, "ImageToFlashcard"));
+
+        //saveFlashcardsFromResponse(response, deck, user.getId()); TODO wait untill CV ready
+
+        bundle.deleteStream(streamedFileLocation);
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: move all bundle related interactions to other place

File: backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                .readEntity(new GenericType<List<FlashcardRepresentation>>() {
+                })
+                .stream()
+                .forEach(flashcardRepresentation -> flashcardDAO.createFlashcard(
+                        flashcardRepresentation.setDeckId(deck.getId()).map()
+                        )
+                );
+    }
+
+    private class CVMessage implements Message {
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: Move to new package. make it public.

File: backend/app/src/test/java/com/bls/patronage/FileIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        client = ClientBuilder.newClient().register(MultiPartFeature.class);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        client.close();
+    }
+
+    @Test
+    @Ignore("This test has Authorization issues, but with real users it should work")
```
Comment (mlotysz) [inline backend/app/src/test/java/com/bls/patronage/FileIntegrationTest.java]: Ignore whole test.

File: backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    @POST
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response uploadFile(
+            @Auth User user,
+            @FormDataParam("file") InputStream uploadedInputStream,
+            @FormDataParam("file") FormDataContentDisposition details) throws Exception {
+
+        final Deck deck = new Deck(UUID.randomUUID(), "foo");
+            final java.nio.file.Path location = baseLocation.resolve(user.getId().toString());
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: Where is logic to determine file path for given account, tile? What is this location? Try to hide details of underlying storage implementation.

File: backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+
+        final URL streamedFileLocation = bundle.persistStream(uploadedInputStream, location.toUri().toURL());
+        final Response response = bundle.informListener(new CVMessage(streamedFileLocation, "ImageToFlashcard"));
+
+        //saveFlashcardsFromResponse(response, deck, user.getId()); TODO wait untill CV ready
```
Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: Uncommented this. it looks you not finished implementation.

Comment (VanDavv) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: I can't read the flashcards when theye're not present

Comment (mlotysz) [inline backend/app/src/main/java/com/bls/patronage/resources/FilesResource.java]: No.  This PR should be ready and wait for real service CV. No code changes should be allowed. Should be tested too. If you need introduce stub _for test_. This comment is *your* bug.

**Comment by mlotysz** [comment]
Create new PR rebased onto master. Learn to use git rebase, please.

**Pull request closed without merging**